File: llm-harness-js/src/stats/llm/llm-stats/final-stage.ts
import { asFiniteNumber, asRecord, type JsonObject } from "../shared.js";

import {
  attachPercentiles,
  blendedPriceValue,
  buildScores,
} from "./scoring.js";
import { type EnrichedRows, type ModelStatsSelectedModel } from "./types.js";

const EMPTY_OPENROUTER_PRICING = {
  weighted_input: null,
  weighted_output: null,
} as const;
const MIN_INTELLIGENCE_COST_TOKEN_THRESHOLD = 1_000_000;
const INTELLIGENCE_COST_TOTAL_COST_KEY = "intelligence_index_cost_total_cost";
const INTELLIGENCE_COST_TOTAL_TOKENS_KEY =
  "intelligence_index_cost_total_tokens";
const NULL_FIELD_PRUNE_THRESHOLD = 0.5;
const NULL_FIELD_PRUNE_RECENT_LOOKBACK_DAYS = 90;
const STABLE_TOP_LEVEL_KEYS = new Set<string>([
  "id",
  "name",
  "provider",
  "logo",
  "attachment",
  "reasoning",
  "release_date",
  "modalities",
  "open_weights",
  "cost",
  "context_window",
  "speed",
  "intelligence",
  "intelligence_index_cost",
  "evaluations",
  "scores",
  "percentiles",
]);

function providerFromId(modelId: unknown): string | null {
  if (typeof modelId !== "string") {
    return null;
  }
  const slashIndex = modelId.indexOf("/");
  if (slashIndex <= 0) {
    return null;
  }
  return modelId.slice(0, slashIndex);
}

function providerFromModel(model: JsonObject): string | null {
  const fromId = providerFromId(model.id);
  if (fromId) {
    return fromId;
  }
  return typeof model.provider_id === "string" ? model.provider_id : null;
}

function buildLogo(model: JsonObject, provider: string | null): string {
  const modelCreator = asRecord(model.model_creator);
  const logoSlug = modelCreator.slug;
  if (typeof logoSlug === "string" && logoSlug.length > 0) {
    return `https://artificialanalysis.ai/img/logos/${logoSlug}_small.svg`;
  }
  return `https://models.dev/logos/${provider ?? "unknown"}.svg`;
}

function buildSpeed(
  model: JsonObject,
  modelId: string | null,
  openRouterSpeedById: Map<string, JsonObject>,
): JsonObject {
  const openRouterSpeed = modelId ? openRouterSpeedById.get(modelId) : null;
  const throughput =
    asFiniteNumber(openRouterSpeed?.throughput_tokens_per_second_median) ??
    asFiniteNumber(model.median_output_tokens_per_second);
  const latency =
    asFiniteNumber(openRouterSpeed?.latency_seconds_median) ??
    asFiniteNumber(model.median_time_to_first_token_seconds);
  const e2eLatency =
    asFiniteNumber(openRouterSpeed?.e2e_latency_seconds_median) ??
    asFiniteNumber(model.median_time_to_first_answer_token) ??
    latency;
  return {
    throughput_tokens_per_second_median: throughput,
    latency_seconds_median: latency,
    e2e_latency_seconds_median: e2eLatency,
  };
}

function buildCost(model: JsonObject, openRouterPricing: JsonObject): unknown {
  const baseCost = asRecord(model.cost);
  const cleanedCost: JsonObject = Object.fromEntries(
    Object.entries(baseCost).filter(([, value]) => value != null),
  );
  const weightedInput = asFiniteNumber(openRouterPricing.weighted_input);
  const weightedOutput = asFiniteNumber(openRouterPricing.weighted_output);
  if (weightedInput != null) {
    cleanedCost.weighted_input = weightedInput;
  }
  if (weightedOutput != null) {
    cleanedCost.weighted_output = weightedOutput;
  }
  const blendedPrice = blendedPriceValue(cleanedCost);
  if (blendedPrice != null) {
    cleanedCost.blended_price = blendedPrice;
  }
  return Object.keys(cleanedCost).length > 0 ? cleanedCost : null;
}

function buildEvaluations(model: JsonObject): unknown {
  const evaluations = asRecord(model.evaluations);
  return Object.keys(evaluations).length > 0 ? evaluations : null;
}

function buildIntelligence(model: JsonObject): unknown {
  const intelligence = asRecord(model.intelligence);
  const nonhallucinationRate = asFiniteNumber(
    intelligence.omniscience_hallucination_rate,
  );
  if (nonhallucinationRate != null) {
    intelligence.omniscience_nonhallucination_rate = nonhallucinationRate;
    delete intelligence.omniscience_hallucination_rate;
  }
  delete intelligence[INTELLIGENCE_COST_TOTAL_COST_KEY];
  delete intelligence[INTELLIGENCE_COST_TOTAL_TOKENS_KEY];
  return Object.keys(intelligence).length > 0 ? intelligence : null;
}

function buildIntelligenceIndexCost(model: JsonObject): unknown {
  const fromRow = asRecord(model.intelligence_index_cost);
  const fromIntelligence = asRecord(model.intelligence);
  const totalCost =
    asFiniteNumber(fromRow.total_cost) ??
    asFiniteNumber(fromIntelligence[INTELLIGENCE_COST_TOTAL_COST_KEY]);
  const totalTokens =
    asFiniteNumber(fromRow.total_tokens) ??
    asFiniteNumber(fromIntelligence[INTELLIGENCE_COST_TOTAL_TOKENS_KEY]);
  const normalized = {
    ...fromRow,
    total_cost: totalCost,
    total_tokens:
      totalTokens != null &&
      totalTokens >= MIN_INTELLIGENCE_COST_TOKEN_THRESHOLD
        ? totalTokens
        : null,
  } as JsonObject;
  const cleaned = Object.fromEntries(
    Object.entries(normalized).filter(([, value]) => value != null),
  );
  return Object.keys(cleaned).length > 0 ? cleaned : null;
}

function intelligencePercentileValue(
  model: ModelStatsSelectedModel,
): number | null {
  return asFiniteNumber(asRecord(model.percentiles).intelligence_percentile);
}

function sortModelsByIntelligencePercentile(
  models: ModelStatsSelectedModel[],
): ModelStatsSelectedModel[] {
  return [...models].sort((left, right) => {
    const leftIntelligence = intelligencePercentileValue(left);
    const rightIntelligence = intelligencePercentileValue(right);
    if (leftIntelligence == null && rightIntelligence == null) {
      return (left.id ?? "").localeCompare(right.id ?? "");
    }
    if (leftIntelligence == null) {
      return 1;
    }
    if (rightIntelligence == null) {
      return -1;
    }
    if (leftIntelligence !== rightIntelligence) {
      return rightIntelligence - leftIntelligence;
    }
    return (left.id ?? "").localeCompare(right.id ?? "");
  });
}

function isPlainObject(value: unknown): value is JsonObject {
  return value != null && typeof value === "object" && !Array.isArray(value);
}

function isWithinRecentLookback(
  releaseDate: string | null,
  lookbackDays: number,
): boolean {
  if (typeof releaseDate !== "string" || releaseDate.length === 0) {
    return false;
  }
  const releaseTimestampMs = Date.parse(releaseDate);
  if (!Number.isFinite(releaseTimestampMs)) {
    return false;
  }
  const cutoffMs = Date.now() - lookbackDays * 24 * 60 * 60 * 1000;
  return releaseTimestampMs >= cutoffMs;
}

function selectPruneSampleModels(
  models: ModelStatsSelectedModel[],
): ModelStatsSelectedModel[] {
  const recentModels = models.filter((model) =>
    isWithinRecentLookback(
      model.release_date,
      NULL_FIELD_PRUNE_RECENT_LOOKBACK_DAYS,
    ),
  );
  return recentModels.length > 0 ? recentModels : models;
}

function countNullishTopLevelKey(
  models: ModelStatsSelectedModel[],
  key: string,
): number {
  return models.reduce((count, model) => {
    const modelRecord = asRecord(model);
    return modelRecord[key] == null ? count + 1 : count;
  }, 0);
}

function countNullishNestedKey(
  models: ModelStatsSelectedModel[],
  parentKey: string,
  nestedKey: string,
): number {
  return models.reduce((count, model) => {
    const modelRecord = asRecord(model);
    const parentValue = modelRecord[parentKey];
    if (!isPlainObject(parentValue) || parentValue[nestedKey] == null) {
      return count + 1;
    }
    return count;
  }, 0);
}

function pruneSparseFields(
  models: ModelStatsSelectedModel[],
  nullThreshold: number = NULL_FIELD_PRUNE_THRESHOLD,
): ModelStatsSelectedModel[] {
  if (models.length === 0) {
    return models;
  }

  const sampleModels = selectPruneSampleModels(models);
  const sampleTotal = sampleModels.length;
  const topLevelKeys = new Set<string>();
  const nestedKeysByParent = new Map<string, Set<string>>();

  for (const model of models) {
    for (const [key, value] of Object.entries(model)) {
      topLevelKeys.add(key);
      if (!isPlainObject(value)) {
        continue;
      }
      const nestedKeys = nestedKeysByParent.get(key) ?? new Set<string>();
      for (const nestedKey of Object.keys(value)) {
        nestedKeys.add(nestedKey);
      }
      nestedKeysByParent.set(key, nestedKeys);
    }
  }

  const topLevelKeysToPrune = new Set<string>();
  for (const key of topLevelKeys) {
    if (STABLE_TOP_LEVEL_KEYS.has(key)) {
      continue;
    }
    const nullCount = countNullishTopLevelKey(sampleModels, key);
    if (nullCount / sampleTotal > nullThreshold) {
      topLevelKeysToPrune.add(key);
    }
  }

  const nestedKeysToPruneByParent = new Map<string, Set<string>>();
  for (const [parentKey, nestedKeys] of nestedKeysByParent) {
    if (parentKey !== "evaluations") {
      continue;
    }
    const keysToPrune = new Set<string>();
    for (const nestedKey of nestedKeys) {
      const nullCount = countNullishNestedKey(
        sampleModels,
        parentKey,
        nestedKey,
      );
      if (nullCount / sampleTotal > nullThreshold) {
        keysToPrune.add(nestedKey);
      }
    }
    if (keysToPrune.size > 0) {
      nestedKeysToPruneByParent.set(parentKey, keysToPrune);
    }
  }

  return models.map((model) => {
    const nextModel: JsonObject = { ...model };
    for (const key of topLevelKeysToPrune) {
      delete nextModel[key];
    }
    for (const [parentKey, nestedKeysToPrune] of nestedKeysToPruneByParent) {
      const parentValue = nextModel[parentKey];
      if (!isPlainObject(parentValue)) {
        continue;
      }
      const nextParentValue: JsonObject = { ...parentValue };
      for (const nestedKey of nestedKeysToPrune) {
        delete nextParentValue[nestedKey];
      }
      nextModel[parentKey] = nextParentValue;
    }
    return nextModel as ModelStatsSelectedModel;
  });
}

function filterModelsById(
  models: ModelStatsSelectedModel[],
  id: string | null | undefined,
): ModelStatsSelectedModel[] {
  if (id == null) {
    return models;
  }
  return models.filter((model) => model.id === id);
}

function projectFinalModel(
  row: unknown,
  openRouterSpeedById: Map<string, JsonObject>,
  openRouterPricingById: Map<string, JsonObject>,
  speedOutputTokenAnchors: number[],
): ModelStatsSelectedModel {
  const model = asRecord(row);
  const provider = providerFromModel(model);
  const modelId = typeof model.id === "string" ? model.id : null;
  const speed = buildSpeed(model, modelId, openRouterSpeedById);
  const pricing =
    (modelId != null ? openRouterPricingById.get(modelId) : null) ??
    EMPTY_OPENROUTER_PRICING;
  const cost = buildCost(model, pricing);
  return {
    id: modelId,
    name: typeof model.name === "string" ? model.name : null,
    provider,
    logo: buildLogo(model, provider),
    attachment: typeof model.attachment === "boolean" ? model.attachment : null,
    reasoning: typeof model.reasoning === "boolean" ? model.reasoning : null,
    release_date:
      typeof model.release_date === "string" ? model.release_date : null,
    modalities: model.modalities ?? null,
    open_weights:
      typeof model.open_weights === "boolean" ? model.open_weights : null,
    cost,
    context_window: model.limit ?? null,
    speed,
    intelligence: buildIntelligence(model),
    intelligence_index_cost: buildIntelligenceIndexCost(model),
    evaluations: buildEvaluations(model),
    scores: buildScores(model, cost, speed, speedOutputTokenAnchors),
    percentiles: null,
  };
}

export function buildFinalModels(
  enrichedRows: EnrichedRows,
  id: string | null | undefined,
): ModelStatsSelectedModel[] {
  const models = enrichedRows.rows.map((row) =>
    projectFinalModel(
      row,
      enrichedRows.openRouterSpeedById,
      enrichedRows.openRouterPricingById,
      enrichedRows.speedOutputTokenAnchors,
    ),
  );
  const modelsWithPercentiles = attachPercentiles(models);
  const sortedModels = sortModelsByIntelligencePercentile(
    modelsWithPercentiles,
  );
  const prunedModels = pruneSparseFields(sortedModels);
  return filterModelsById(prunedModels, id);
}

File: llm_harness/stats/llm/llm_stats_stages/final_stage.py
"""Final projection stage for native Python LLM stats."""

from __future__ import annotations

from typing import Any

from ..shared import as_finite_number, as_record
from .scoring import attach_relative_scores, blended_price_value, build_scores

EMPTY_OPENROUTER_PRICING = {"weighted_input": None, "weighted_output": None}
MIN_INTELLIGENCE_COST_TOKEN_THRESHOLD = 1_000_000
INTELLIGENCE_COST_TOTAL_COST_KEY = "intelligence_index_cost_total_cost"
INTELLIGENCE_COST_TOTAL_TOKENS_KEY = "intelligence_index_cost_total_tokens"


def _provider_from_id(model_id: Any) -> str | None:
    """Resolve the provider for Final-stage LLM stats selection."""
    if not isinstance(model_id, str) or "/" not in model_id:
        return None
    return model_id.split("/", 1)[0]


def _provider_from_model(model: dict[str, Any]) -> str | None:
    """Resolve the provider for Final-stage LLM stats selection."""
    return _provider_from_id(model.get("id")) or (model.get("provider_id") if isinstance(model.get("provider_id"), str) else None)


def _build_logo(model: dict[str, Any], provider: str | None) -> str:
    """Build the logo field for Final-stage LLM stats selection."""
    if isinstance(model.get("logo"), str) and model.get("logo"):
        return model["logo"]
    logo_slug = as_record(model.get("model_creator")).get("slug")
    if isinstance(logo_slug, str) and logo_slug:
        return f"https://artificialanalysis.ai/img/logos/{logo_slug}_small.svg"
    return f"https://models.dev/logos/{provider or 'unknown'}.svg"


def _build_speed(model: dict[str, Any], model_id: str | None, openrouter_speed_by_id: dict[str, dict[str, Any]]) -> dict[str, Any]:
    """Build the speed score for Final-stage LLM stats selection."""
    openrouter_speed = openrouter_speed_by_id.get(model_id, {}) if model_id else {}
    throughput = as_finite_number(openrouter_speed.get("throughput_tokens_per_second_median")) or as_finite_number(model.get("median_output_tokens_per_second"))
    latency = as_finite_number(openrouter_speed.get("latency_seconds_median")) or as_finite_number(model.get("median_time_to_first_token_seconds"))
    e2e_latency = as_finite_number(openrouter_speed.get("e2e_latency_seconds_median")) or as_finite_number(model.get("median_time_to_first_answer_token")) or latency
    return {
        "throughput_tokens_per_second_median": throughput,
        "latency_seconds_median": latency,
        "e2e_latency_seconds_median": e2e_latency,
    }


def _build_cost(model: dict[str, Any], openrouter_pricing: dict[str, Any], scoring_config: dict[str, Any]) -> dict[str, Any] | None:
    """Build the cost score for Final-stage LLM stats selection."""
    base_cost = as_record(model.get("cost"))
    cleaned_cost = {key: value for key, value in base_cost.items() if value is not None}
    weighted_input = as_finite_number(openrouter_pricing.get("weighted_input"))
    weighted_output = as_finite_number(openrouter_pricing.get("weighted_output"))
    if weighted_input is not None:
        cleaned_cost["weighted_input"] = weighted_input
    if weighted_output is not None:
        cleaned_cost["weighted_output"] = weighted_output
    blended_price = blended_price_value(cleaned_cost, scoring_config)
    if blended_price is not None:
        cleaned_cost["blended_price"] = blended_price
    return cleaned_cost or None


def _build_intelligence(model: dict[str, Any]) -> dict[str, Any] | None:
    """Build the intelligence score for Final-stage LLM stats selection."""
    intelligence = dict(as_record(model.get("intelligence")))
    nonhallucination_rate = as_finite_number(intelligence.get("omniscience_hallucination_rate"))
    if nonhallucination_rate is not None:
        intelligence["omniscience_nonhallucination_rate"] = nonhallucination_rate
        intelligence.pop("omniscience_hallucination_rate", None)
    intelligence.pop(INTELLIGENCE_COST_TOTAL_COST_KEY, None)
    intelligence.pop(INTELLIGENCE_COST_TOTAL_TOKENS_KEY, None)
    return intelligence or None


def _build_intelligence_index_cost(model: dict[str, Any]) -> dict[str, Any] | None:
    """Build the intelligence index cost for Final-stage LLM stats selection."""
    from_row = as_record(model.get("intelligence_index_cost"))
    from_intelligence = as_record(model.get("intelligence"))
    total_cost = as_finite_number(from_row.get("total_cost")) or as_finite_number(from_intelligence.get(INTELLIGENCE_COST_TOTAL_COST_KEY))
    total_tokens = as_finite_number(from_row.get("total_tokens")) or as_finite_number(from_intelligence.get(INTELLIGENCE_COST_TOTAL_TOKENS_KEY))
    normalized = {
        **from_row,
        "total_cost": total_cost,
        "total_tokens": total_tokens if total_tokens is not None and total_tokens >= MIN_INTELLIGENCE_COST_TOKEN_THRESHOLD else None,
    }
    cleaned = {key: value for key, value in normalized.items() if value is not None}
    return cleaned or None


def _project_final_model(
    row: dict[str, Any],
    openrouter_speed_by_id: dict[str, dict[str, Any]],
    openrouter_pricing_by_id: dict[str, dict[str, Any]],
    speed_output_token_anchors: list[int],
    scoring_config: dict[str, Any],
) -> dict[str, Any]:
    """Helper for project final model."""
    model = as_record(row)
    provider = _provider_from_model(model)
    model_id = model.get("id") if isinstance(model.get("id"), str) else None
    speed = _build_speed(model, model_id, openrouter_speed_by_id)
    pricing = openrouter_pricing_by_id.get(model_id, EMPTY_OPENROUTER_PRICING) if model_id else EMPTY_OPENROUTER_PRICING
    cost = _build_cost(model, pricing, scoring_config)
    return {
        "id": model_id,
        "name": model.get("name") if isinstance(model.get("name"), str) else None,
        "provider": provider,
        "logo": _build_logo(model, provider),
        "attachment": model.get("attachment") if isinstance(model.get("attachment"), bool) else None,
        "reasoning": model.get("reasoning") if isinstance(model.get("reasoning"), bool) else None,
        "release_date": model.get("release_date") if isinstance(model.get("release_date"), str) else None,
        "modalities": model.get("modalities"),
        "open_weights": model.get("open_weights") if isinstance(model.get("open_weights"), bool) else None,
        "cost": cost,
        "context_window": model.get("limit"),
        "speed": speed,
        "intelligence": _build_intelligence(model),
        "intelligence_index_cost": _build_intelligence_index_cost(model),
        "evaluations": as_record(model.get("evaluations")) or None,
        "scores": build_scores(model, cost, speed, speed_output_token_anchors, scoring_config),
        "relative_scores": None,
    }


def build_final_payload(
    rows: list[dict[str, Any]],
    *,
    model_id: str | None = None,
    fetched_at_epoch_seconds: int | None = None,
    openrouter_speed_by_id: dict[str, dict[str, Any]] | None = None,
    openrouter_pricing_by_id: dict[str, dict[str, Any]] | None = None,
    speed_output_token_anchors: list[int] | None = None,
    scoring_config: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Build the final Final-stage LLM stats selection payload."""
    openrouter_speed_by_id = openrouter_speed_by_id or {}
    openrouter_pricing_by_id = openrouter_pricing_by_id or {}
    speed_output_token_anchors = speed_output_token_anchors or [200, 500, 1000, 2000, 8000]
    scoring_config = scoring_config or {}
    models = [
        _project_final_model(
            row,
            openrouter_speed_by_id,
            openrouter_pricing_by_id,
            speed_output_token_anchors,
            scoring_config,
        )
        for row in rows
    ]
    models = attach_relative_scores(models)
    models.sort(
        key=lambda model: (
            -(
                as_finite_number(as_record(model.get("relative_scores")).get("intelligence_score"))
                if as_finite_number(as_record(model.get("relative_scores")).get("intelligence_score")) is not None
                else float("-inf")
            ),
            model.get("id") or "",
        )
    )
    if model_id is not None:
        models = [row for row in models if row.get("id") == model_id]
    return {
        "fetched_at_epoch_seconds": fetched_at_epoch_seconds,
        "models": models,
    }
